dotnet/src/SemanticKernel/Planning/Plan.cs

```diff
@@ -470,21 +470,49 @@ private SKContext UpdateContextWithOutputs(SKContext context)
     /// <returns>The context variables for the next step in the plan.</returns>
     private ContextVariables GetNextStepVariables(ContextVariables variables, Plan step)
     {
-        // If the current step is passing to another plan, we set the default input to an empty string.
-        // Otherwise, we use the description from the current plan as the default input.
-        // We then set the input to the value from the SKContext, or the input from the Plan.State, or the default input.
-        var defaultInput = step.Steps.Count > 0 ? string.Empty : this.Description ?? string.Empty;
-        var planInput = string.IsNullOrEmpty(variables.Input) ? this.State.Input : variables.Input;
-        var stepInput = string.IsNullOrEmpty(planInput) ? defaultInput : planInput;
-        var stepVariables = new ContextVariables(stepInput);
+        // Priority for Input
+        // - Parameters (expand from variables if needed)
+        // - SKContext.Variables
+        // - Plan.State
+        // - Empty if sending to another plan
+        // - Plan.Description
+
+        var input = string.Empty;
+        if (!string.IsNullOrEmpty(step.Parameters.Input))
+        {
+            input = this.ExpandFromVariables(variables, step.Parameters.Input);
+        }
+        else if (!string.IsNullOrEmpty(variables.Input))
+        {
+            input = variables.Input;
+        }
+        else if (!string.IsNullOrEmpty(this.State.Input))
+        {
+            input = this.State.Input;
+        }
+        else if (step.Steps.Count > 0)
+        {
+            input = string.Empty;
+        }
+        else if (!string.IsNullOrEmpty(this.Description))
+        {
+            input = this.Description;
+        }
+
+        var stepVariables = new ContextVariables(input);
+
 
         // Priority for remaining stepVariables is:
-        // - Parameters (pull from State by a key value)
-        // - Parameters (from context)
-        // - Parameters (from State)
+        // - Function Parameters (pull from variables or state by a key value)
+        // - Step Parameters (pull from variables or state by a key value)
         var functionParameters = step.Describe();
         foreach (var param in functionParameters.Parameters)
         {
+            if (param.Name.Equals("INPUT", StringComparison.OrdinalIgnoreCase))
+            {
+                continue;
+            }
+
             if (variables.Get(param.Name, out var value) && !string.IsNullOrEmpty(value))
```

Comment:
```suggestion
if (!string.IsNullOrEmpty(value) && 
    (variables.Get(param.Name, out var value) 
    || this.State.Get(param.Name, out value)))
{
    stepVariables.Set(param.Name, value);
}

```

Reply:
Same for the others... check for empty/null first.

Reply:
`Get` could technically return an empty string -- I don't think it should then be used.

Reply:
Okay, I'm going back on this, thanks. If it's an empty string, something did that on purpose, let's respect it.